apis/storage/v1alpha3/bucketpolicy_types.go

```diff
@@ -0,0 +1,197 @@
+		m["Id"] = p.PolicyID
+	}
+	slc := make([]interface{}, len(p.PolicyStatement))
+	for i, v := range p.PolicyStatement {
+		msg, err := v.Serialize()
+		if err != nil {
+			return nil, err
+		}
+		slc[i] = msg
+	}
+	m["Statement"] = slc
+	//dat, err := json.Marshal(m)
+	//if err != nil {
+	//	return nil, err
+	//}
+	//str := string(dat)
+	return m, nil
+}
+
+// S3BucketPolicyStatement defines an individual statement within the
+// S3BucketPolicyBody
+type S3BucketPolicyStatement struct {
+	// Optional identifier for this statement, must be unique within the
+	// policy if provided.
+	StatementID string `json:"Sid,omitempty"`
+
+	// The effect is required and specifies whether the statement results
+	// in an allow or an explicit deny. Valid values for Effect are Allow and Deny.
+	Effect string `json:"Effect"`
+
+	// Used with the S3 policy to specify the principal that is allowed
+	// or denied access to a resource.
+	Principal *S3BucketPrincipal `json:"Principal,omitempty"`
+
+	// Each element of the PolicyAction array describes describes the specific
+	// action or actions that will be allowed or denied with this PolicyStatement.
+	PolicyAction []string `json:"Action"`
+
+	// This flag indicates that this policy should apply to the IAMUsername
+	// that was either passed in or created for this bucket.
+	ApplyToIAMUser bool `json:"EffectIAMUser,omitempty"`
+
+	// The paths on which this resource will apply
+	ResourcePath []string `json:"Resource"`
+}
+
+//Serialize is the custom marshaller for the S3BucketPolicyStatement
+func (p *S3BucketPolicyStatement) Serialize() (interface{}, error) {
+	m := make(map[string]interface{})
+	principal, err := p.Principal.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	m["Principal"] = principal
+	m["Action"] = tryFirst(p.PolicyAction)
+	m["Resource"] = tryFirst(p.ResourcePath)
+	m["Effect"] = p.Effect
+	if p.StatementID != "" {
+		m["Sid"] = p.StatementID
+	}
+	return m, nil
+}
+
+// S3BucketPrincipal defines the principal users affected by
+// the S3BucketPolicyStatement
+type S3BucketPrincipal struct {
+	// This flag indicates if the policy should be made available
+	// to all anonymous users.
+	AllowAnon bool `json:"AllowAnon,omitempty"`
+
+	// This list contains the all of the AWS IAM users which are affected
+	// by the policy statement
+	AWSPrincipal []string `json:"AWS,omitempty"`
+}
+
+func tryFirst(slc []string) interface{} {
+	if len(slc) == 1 {
+		return slc[0]
+	}
+	return slc
+}
+
+//Serialize is the custom serializer for the S3BucketPrincipal
+func (p *S3BucketPrincipal) Serialize() (interface{}, error) {
+	all := "*"
+	if p.AllowAnon {
+		return all, nil
+	}
+	m := make(map[string]interface{})
+	m["AWS"] = tryFirst(p.AWSPrincipal)
+	return m, nil
+}
+
+// An S3BucketPolicySpec defines the desired state of an
+// S3BucketPolicy.
+type S3BucketPolicySpec struct {
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+	PolicyBody                   S3BucketPolicyParameters `json:"forProvider"`
```

Comment:
Typically all fields that are part of building the external state for the Bucket should be part of `forProvider` (including `bucketName` and `userName` fields below)